rust/main/chains/hyperlane-sealevel/src/provider.rs

```diff
@@ -189,7 +192,7 @@ impl HyperlaneProvider for SealevelProvider {
             gas_price,
             nonce: 0,
             sender,
-            recipient: None,
+            recipient: Some(self.contract_address),
```

Comment:
for my own understanding - bc there can be multiple instructions in Sealevel, you didn't want to grab the recipient of one of the instructions in tx (because it's not clear which instruction we care about), and instead would prefer to always have the Mailbox address?

what about txs that aren't made to the Mailbox, e.g. for message dispatches or gas payments?

Fwiw this field isn't really used so I don't think the correctness of it matters much, but I could see this being a bit confusing having every recipient be the Mailbox even when it's not the case

Reply:
There can be multiple instructions in a Sealevel transaction on the highest level and they can be Native Programs or any other programs. To reliably identify a program which eventually invokes the Mailbox (so, the program is the proper recipient of the transaction), we need to simulate the execution of this program, I think.

Message dispatches are made by Mailbox program, as far as I can tell. Gas payments are done by a separate program, so, it is lost in this implementation.

We can discuss if there are easier ways to identify the recipient of the transaction.

Reply:
How about checking if `transaction.message.account_keys.contains(self.contract_address)` to gain more confidence in setting `recipient` to `Some`, and otherwise set it to `None`? Because even if we use this provider method in a superfluous way, it could theoretically be used to fetch any tx whatsoever by hash

Reply:
actually I think `tx_with_meta.meta.loaded_addresses.writable.contains(...)` would be even more robust

Reply:
Discussed in person. We extract program from the top-level instructions.